### PR TITLE
Add system.runtime.nodes information to web UI

### PR DESCRIPTION
**Pull request by whutpencil**
## Description

The fact that you can't see a list of workers in the Web UI doesn't seem elegant. The goal of this PR is to improve it. The effect is shown in the figure below:

image

The value of active workers **became a hyperlink.**

When you click it, you will enter a new page with the contents as you run `select * from system.runtime.nodes;` the results obtained.

At the same time, each node ID and node IP supports jumping to the corresponding machine to view information.

image

image

image

This is the effect of having multiple nodes in a formal cluster:

image

> Is this change a fix, improvement, new feature, refactoring, or other?
Improvement.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
No.

> How would you describe this change to a non-technical end user or system administrator?

## Documentation

(x) No documentation is needed.
( ) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(x) Release notes entries required with the following suggested text:

```markdown
# Web UI
* Add list of nodes to UI. ({issue}`11653`)
```

### CODE REVIEW

**Comment by findepi** [comment]
@dedep PTAL

**Comment by whutpencil** [comment]
cc @dedep

**Comment by whutpencil** [comment]
@findepi Why hasn't anyone reviewed the code for so long? What happened to Trino community activity? If I can, I am willing to participate in the review process of assisting the community. LOL ^^

**Comment by findepi** [comment]
@whutpencil i am not a frontend dev person, so I wouldn't want to review myself, but let me try again find someone.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        clearTimeout(this.timeoutId); // to stop multiple series of refreshLoop from going on simultaneously
+        // const nodeId = getFirstParameter(window.location.search);
+        $.get('/ui/api/worker', function (workerInfo) {
+            this.setState({
+                initialized: true,
+                workerInfo: workerInfo
+            })
+            for (var index in workerInfo) {
+                this.setState({
+                    workerId: addToHistory(workerInfo[index].nodeId, this.state.workerId),
```
Comment (dedep) [inline core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:58]: `addToHistory` function signature is:
`export function addToHistory (value: number, valuesArray: number[]): number[] {`

You don't pass numbers, but strings.
I don't think you should use this function here.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    refreshLoop() {
+        clearTimeout(this.timeoutId); // to stop multiple series of refreshLoop from going on simultaneously
+        // const nodeId = getFirstParameter(window.location.search);
+        $.get('/ui/api/worker', function (workerInfo) {
+            this.setState({
+                initialized: true,
+                workerInfo: workerInfo
+            })
+            for (var index in workerInfo) {
+                this.setState({
```
Comment (dedep) [inline core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:57]: I think that you didn't design the state well. Instead of
```
{
            workerId: string[],
            workerIp: string[],
            workerVersion: string[],
            ...
}
```
I would do (typescript notation):
```
{
   workers: [
            workerId: string,
            workerIp: string,
            workerVersion: string,
            ...
   ]
}
```

This way, the model is more concise and easier to work on

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java [hunk 3]
```diff
@@ -123,6 +127,78 @@ public Response getThreads(
...
+            return coordinator;
+        }
+
+        @JsonProperty
+        public String getState()
+        {
+            return state;
+        }
+    }
+
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java:203]: done.

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java [hunk 4]
```diff
@@ -123,6 +127,78 @@ public Response getThreads(
...
+        private final String state;
+
+        @JsonCreator
+        public JsonNodeInfo(@JsonProperty("nodeId") String nodeId,
+                @JsonProperty("nodeIp") String nodeIp,
+                @JsonProperty("nodeVersion") String nodeVersion,
+                @JsonProperty("coordinator") String coordinator,
+                @JsonProperty("state") String state)
+        {
+            this.nodeId = nodeId;
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java]: done.

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java [hunk 5]
```diff
@@ -123,6 +127,78 @@ public Response getThreads(
...
+        return Response.ok().entity(jsonNodes).build();
+    }
+
+    public static class JsonNodeInfo
+    {
+        private final String nodeId;
+        private final String nodeIp;
+        private final String nodeVersion;
+        private final String coordinator;
+        private final String state;
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java:157]: done.

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java [hunk 6]
```diff
@@ -123,6 +127,78 @@ public Response getThreads(
...
+        }
+        return Response.ok().entity(jsonNodes).build();
+    }
+
+    public static class JsonNodeInfo
+    {
+        private final String nodeId;
+        private final String nodeIp;
+        private final String nodeVersion;
+        private final String coordinator;
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java]: done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            workerVersion: [],
+            coordinator: [],
+            state: [],
+        };
+        this.refreshLoop = this.refreshLoop.bind(this);
+    }
+
+    refreshLoop() {
+        clearTimeout(this.timeoutId); // to stop multiple series of refreshLoop from going on simultaneously
+        // const nodeId = getFirstParameter(window.location.search);
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+                    state: addToHistory(workerInfo[index].state, this.state.state)
+                });
+            }
+        }.bind(this))
+                .fail(() => {
+                    this.setState({
+                        initialized: true,
+                    });
+                });
+
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:61]: done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+            workerIp: [],
+            workerVersion: [],
+            coordinator: [],
+            state: [],
+        };
+        this.refreshLoop = this.refreshLoop.bind(this);
+    }
+
+    refreshLoop() {
+        clearTimeout(this.timeoutId); // to stop multiple series of refreshLoop from going on simultaneously
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: done.

File: core/trino-main/src/main/resources/webapp/workers.html [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+<body>
+
+<div class="container">
+    <div id="title"></div>
+
+    <div id="worker-list">
+        <div class="loader">Loading...</div>
+    </div>
+    <!--<div id="worker-threads">
```
[the remark this answers is not in the document]
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/workers.html:48]: done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx (hunk 1 above)
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx (hunk 2 above)
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: done.

**Comment by whutpencil** [comment]
@dedep Thank you very much for your detailed review. I'm not a full-time front-end staff, so I didn't use the corresponding data structure well in the implementation process. I learned a lot from your comments and have made changes according to your modification opinions. Please review again.

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java [hunk 11]
```diff
@@ -123,6 +130,77 @@ public Response getThreads(
...
+        public JsonNodeInfo(@JsonProperty("nodeId") String nodeId,
+                @JsonProperty("nodeIp") String nodeIp,
+                @JsonProperty("nodeVersion") String nodeVersion,
+                @JsonProperty("coordinator") boolean coordinator,
+                @JsonProperty("state") String state)
+        {
+            this.nodeId = requireNonNull(nodeId, "nodeId is null");
+            this.nodeIp = requireNonNull(nodeIp, "nodeIp is null");
+            this.nodeVersion = requireNonNull(nodeVersion, "nodeVersion is null");
+            this.coordinator = requireNonNull(coordinator, "coordinator is null");
```
Comment (dedep) [inline core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java:169]: There is no need to do the `requireNonNull` check for a primitive (`boolean`) type

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+};
+
+export class WorkerList extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            workerInfo: null,
+            initialized: false,
+            ended: false,
+            workers: [{
```
Comment (dedep) [inline core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:34]:
```
                workerId: String,
                workerIp: String,
                workerVersion: String,
                coordinator: Boolean,
                state: String
```

is not a correct JavaScript code. The code snippet I posted before was written in Typescript to show you the types.

So, I think it should be just: `workers: []`

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                coordinator: Boolean,
+                state: String
+            }]
+        };
+        this.refreshLoop = this.refreshLoop.bind(this);
+    }
+
+    refreshLoop() {
+        clearTimeout(this.timeoutId);
+        $.get('/ui/api/worker', function (workerInfo) {
```
Comment (dedep) [inline core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:47]: I would rename `workerInfo` to `workers` to signal that this is an array.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+    disableHiddenCheck: true,
+};
+
+export class WorkerList extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = {
+            workerInfo: null,
+            initialized: false,
+            ended: false,
```
Comment (dedep) [inline core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx:33]: I think that `ended` property is not used anywhere. IMO, you can remove this line

File: core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java (hunk 11 above)
Comment (whutpencil) [reply core/trino-main/src/main/java/io/trino/server/ui/WorkerResource.java]: Done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx (hunk 12 above)
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: Done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx (hunk 13 above)
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: Done.

File: core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx (hunk 14 above)
Comment (whutpencil) [reply core/trino-main/src/main/resources/webapp/src/components/WorkerList.jsx]: Done.

**Comment by whutpencil** [comment]
@dedep Excuse me, do you have any further amendments?  Could this PR be merged into the master?

**Comment by electrum** [comment]
Thanks for your contribution!

**Comment by mosabua** [comment]
Hey @whutpencil  .. I would like to feature this PR in the Trino Community Broadcast next week. Would you be interested in joining me a guest in the episode?

**Comment by mosabua** [comment]
I am going ahead with demoing this in Trino Community Broadcast 35 on Thursday ..

**Comment by whutpencil** [comment]
font{
        line-height: 1.6;
    }
    ul,ol{
        padding-left: 20px;
        list-style-position: inside;
    }


    
    
        Hello, I'm very sorry that I'm busy recently due to my family. I didn't pay attention to the email and reply to your message in time. I am honored to contribute to Trino community. Thank you very much for your invitation. Do I need to prepare anything?
            
        
    
    
        
            
        
    
    
      
    
        
                    
            
                
                    
                            
                                
                            
                            
                                137497191
                            
                    
                        
                            
                                    ***@***.***
                            
                        
                
            
        
        
    
    
        签名由
        网易邮箱大师
        定制
    
 
    

    On 04/20/2022 14:43，Manfred ***@***.***> wrote： 



I am going ahead with demoing this in Trino Community Broadcast 35 on Thursday ..

—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you were mentioned.Message ID: ***@***.***>

**Comment by mosabua** [comment]
You can watch the demo and such now.. https://trino.io/episodes/35.html
